Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,72 @@
+# Spectrum Number and Workspace Index Abstraction Layer
+
+
+## Motivation
+
+There are two main motivations for this design proposal:
+
+1. We need to provide a consistent user interface that allows users to specify which spectrums in a workspace an algorithm should modify.
+2. Developers of algorithms should not have to deal with conversions between spectrum number and other index types.
+
+
+## Current situation and future problems
+
+#### Definitions:
+1. Workspace index = linear index starting at zero, corresponds to memory offset in vector of spectra
+2. Spectrum number = user defined index for spectra in certain workspace
+
+#### Current situation:
+
+- Conversion methods between index types (spectrum number and workspace index) are part of the workspace interface.
+- There is no consistent way of defining index ranges or lists for algorithms. This is done individually in each algorithm, and properties may be defined for spectrum number or workspace indices, and sometimes the interpretation of index properties even seems to depend on the facility.
+- Validation of index ranges and behavior on errors such as out-of-range indices is likewise done individually in all algorithms. Algorithms behave inconsistently depending on errors (throw and error, print a warning, ignore silently). There are bugs in some range validations.
+
+#### Future problems:
+
+When running Mantid with MPI we will most likely store different spectra on different MPI ranks.
+As a consequence the workspace index will not longer correspond to the memory offset in the local vector of spectra.
+Basically we will have a global index and a local index on each rank.
+The local indices from all ranks are disjoint sets that together make up all global indices.
+
+This introduces additional complexity in the index handling and will require definition of new terms.
+
+
+## Detailed goals
+
+- All algorithms should have a consistent interface.
+- Eliminate duplicated code for declaring properties for index ranges and list.
+- Eliminate duplicated code for validating indices and converting them to the desired type.
+- Developers should generally not have to deal with spectrum number at all. Spectrum number is important for the user interface, but irrelevant for the internals of most algorithms.
+- Developers should in the future be isolated from dealing explicitly with the local and global indices that come with an MPI implementation of Mantid.
+
+
+## User interface
+
+- The interface of algorithms should make it possible to specify spectrum number ranges and lists.
+- For better compatibility for the current interface and for users who do not like to deal with spectrum number it should also be possible to specify workspace index ranges and lists.
+
```

Comment:
If the user interface side of things are to be fixed. I think it would make sense to introduce a property type for spectrum ranges. At the moment either `PropertyWithValue<int>` or `ArrayProperty<int>` are used for such purposes, developers then usually have to to use a `BoundedValidator` to prevent < 0 being used. Developers are currently then responsible for specifying what the input actually means (workspace indexes, others as spectrum numbers and others as detector ids). I think that a new property type that encapsulates all of this could be a solution to the UI side of the problem.